serialization: c++20 endian/byteswap/clz modernization

theuni opened the pull request:
This replaces #28674, #29036, and #29057. Now ready for testing and review.

Replaces platform-specific endian and byteswap functions. This is especially useful for kernel, as it means that our deep serialization code no longer requires bitcoin-config.h.

I apologize for the size of the last commit, but it's hard to avoid making those changes at once.

All platforms now use our internal functions rather than libc or platform-specific ones, with the exception of MSVC.

Sadly, benchmarking showed that not all compilers are capable of detecting and optimizing byteswap functions, so compiler builtins are instead used where possible. However, they're now detected via macros rather than autoconf checks.

This[ matches how libc++ implements std::byteswap for c++23](https://github.com/llvm/llvm-project/blob/main/libcxx/include/__bit/byteswap.h#L26).

I suggest we move/rename `compat/endian.h`, but I left that out of this PR to avoid bikeshedding.

#29057 pointed out some irregularities in benchmarks. After messing with various compilers and configs for a few weeks with these changes, I'm of the opinion that we can't win on every platform every time, so we should take the code that makes sense going forward. That said, if any real-world slowdowns are caused here, we should obviously investigate.

theuni committed:
crypto: replace non-standard CLZ builtins with c++20's bit_width

Also some header cleanups.

theuni committed:
crypto: replace CountBits with std::bit_width

bit_width is a drop-in replacement with an exact meaning in c++, so there is
no need to continue testing/fuzzing/benchmarking.

theuni commented:
Here's a godbolt test that shows how various compilers perform: https://gcc.godbolt.org/z/nTadqEP83

To test with/without builtins, comment/un-comment the `DISABLE_BUILTIN_BSWAPS` define at line 14.

From my tests:
- clang: all c++20 supporting versions (>= 10.0) optimize without the need for builtins
- gcc: no version optimizes without the builtins
- msvc: versions >= 19.37 optimize without builtins

theuni commented:
Added a quick note about `std::byteswap` and c++23.

fanquake commented:
Concept ACK - I think moving to using the builtins is ok. With the eventual plan to migrate to `std::byteswap`.

@aureleoules everytime I visit the [coverage/benchmarks for this PR](https://corecheck.dev/bitcoin/bitcoin/pulls/29263), I get a 500 error. Can you take a look?

aureleoules commented:
@fanquake thanks for reaching out, I fixed the issue.

theuni commented:
@aureleoules Thanks! These numbers really don't make much sense to me and don't match what I've seen locally, but I'll work on trying to reproduce.

theuni commented:
I've dropped the clz changes as a test and kept only the endian/byteswap change. Locally on my machine the benchmarks look the same before and after.

If the corecheck benchmarks look better I'll update the title/description here.

theuni commented:
Aha, I finally tracked down the culprit! No wonder the benchmarks weren't making sense!

The problem was the removal of this from `crypto/common.h`:
```c++
#if defined(HAVE_CONFIG_H)
#include <config/bitcoin-config.h>
#endif
```

Turns out, `sha256.cpp` was relying on that. Adding that include to `sha256.cpp` (where it belongs) fixes my local benchmarks.

I'm going to put the clz changes back to match the PR title/description, but this time with the include fixed up. Hopefully after that the benchmarks will make sense and all will be good :)

I suppose there might be other compilation units that were depending on the include indirectly as well. I'll figure out a way to check for that.

theuni commented:
Converted to a draft while I'm still messing with this.

I added a new commit to add the `bitcoin-config.h` include where it's necessary. I'm not done investigating that yet, but it's at least needed for `chacha20poly1305.cpp` as well.

The benchmarks still show some regressions, though it's not nearly as bad as it was before. Will continue poking.

theuni commented:
I've pushed a commit which temporarily puts the includes back in the low-level headers for the sake of addressing the benchmarks first, setting aside the possibility of missing defines.

@aureleoules's [benchmarks show 2 small regressions](https://corecheck.dev/bitcoin/bitcoin/pulls/29263), though I am unable to reproduce those locally.

My tests show: `./bench/bench_bitcoin -min-time=5000 -filter="AddrManGetAddr|PrevectorDeserializeNontrivial"`
master (3d52cedb497e0ec029ba3cef95b00169c157d8ae):
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          224,138.83 |            4,461.52 |    0.1% |      5.50 | `AddrManGetAddr`
|              155.96 |        6,411,753.11 |    0.1% |      5.49 | `PrevectorDeserializeNontrivial
```
This PR:
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          226,531.50 |            4,414.40 |    0.2% |      5.50 | `AddrManGetAddr`
|              156.91 |        6,372,901.08 |    0.1% |      5.50 | `PrevectorDeserializeNontrivial`
```

Essentially no change. I'm curious if @maflcko sees the same?

aureleoules commented:
@theuni yeah the `AddrManGetAddr` and `PrevectorDeserializeNontrivial` are known to be flaky. I'll filter them out from the UI while I find a better solution.
Some benchmarks depend on I/O or randomness so it's hard a have consistent results between runs.

You can see ignored benchmarks here in the meantime: https://github.com/corecheck/frontend/blob/master/src/routes/%5Bowner%5D/%5Brepo%5D/pulls/%5Bnumber%5D/Benchmarks.svelte#L334

theuni commented:
@aureleoules Thanks, that's helpful.

fanquake commented:
@theuni want to rebase this now that #29404 is in?

theuni commented:
Rebased. We'll see what c-i thinks now. Benchmarks should be all good.

I think it makes sense for #29408 to go first though.

theuni commented:
Rebased after #29408 merge and undrafted.

theuni commented on src/compat/byteswap.h:
```diff
@@ -1,59 +0,0 @@
-          | ((x & 0x00000000ff000000ull) << 8)
-          | ((x & 0x0000000000ff0000ull) << 24)
-          | ((x & 0x000000000000ff00ull) << 40)
-          | ((x & 0x00000000000000ffull) << 56));
```
Hmm, I don't remember now honestly. My best guess is I thought this would make the change to `std::byteswap` more straightforward in the future, but agree that it'd be easier to review now if left in-place. I'll move it back.

theuni replied:
Ah, I guess I combined them because the `constexpr`-ness of the endian conversion functions is at the mercy of the byteswap functions, so `compat/endian.h` requires the use of the `BSWAP_CONSTEXPR` macro. I guess that's still ok with them split up, though.

theuni replied:
No, but I figure compilers are more likely to be able to optimize in the non-builtin case if they're constexpr.

theuni replied:
Indeed, thanks. The whitespace was already off on the first line, so I kept it that way to reduce the diff here.

theuni committed:
serialization: detect byteswap builtins without autoconf tests

Rather than a complicated set of tests to decide which bswap functions to
use, always prefer the compiler built-ins when available.

These builtins and fallbacks can all be removed once we're using c++23, which
adds std::byteswap.

theuni committed:
serialization: use internal endian conversion functions

These replace our platform-specific mess in favor of c++20 endian detection
via std::endian and internal byteswap functions when necessary.

They no longer rely on autoconf detection.

maflcko reviewed (commented):
ACK 1d051e86098e721a47cebf7cfd73d82aa414c05e 📛

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: ACK 1d051e86098e721a47cebf7cfd73d82aa414c05e 📛
6Obp6pvdrMxN7MXEPLAuvZe0UCy9RzNtRu/VNpLvSGe6P6thBW12J6z7gacvZBa1MGOdmgJFjbeCHc8MjWTyDA==
```

</details>

theuni commented on src/compat/byteswap.h:
```diff
@@ -5,44 +5,67 @@
 
-#if defined(HAVE_CONFIG_H)
-#include <config/bitcoin-config.h>
+#include <bit>
```
Nice catch, this was leftover from the combined commit. Removed.

fanquake reviewed (approved):
ACK 86b7f28d6c507155a9d3a15487ee883989b88943 - we can finish pruning out the __builtin_clz* checks/usage once the minisketch code has been updated. This is more good cleanup pre-CMake & for the kernal.
